### PR TITLE
Msgpack serializer

### PR DESCRIPTION
**Pull request by danmarcab**
@jtmoulia, hopefully I haven't committed sacrilege with this change. I had to tweak also the WebSocket transport a bit to support different frame types(:text or :binary). This is my first public commit in elixir so feel free to point out any possible improvements.

I think it would be nice to run the integration tests with all the serializers (and maybe in the future with all the transports), just to make sure all are working for all use cases. 
Do you have any ideas on how to implement that? I don't want to duplicate every single test.

### CODE REVIEW

**Comment by jtmoulia** [comment]
Hey @danmarcab -- looks great!

Separate but related to your PR: I started us down the path of exposing transport specific information for serializers using callbacks with `name`, and you're right -- we additionally need to parameterize `frame_type` to suport both JSON + msgpack. But, the serializer callback interface is going to get ugly if it's necessary to keep adding callbacks to support transport specific functionality.

Alternatives which keep the interface slim are:

To continue with  `Spell.Serializer`, all of the transport info could be returned by a single callback function:

```elixir
defcallback transport_info(module :: atom) :: map
```

An implementation for msgpack might look like:

```elixir
def transport_info(Spell.Transport.Websocket) do
  %{name: "msgpack", frame_type: :binary}
end
```

Alternatively, similar to the WAMP basic protocol docs, the transport could be aware of the serializers. In `Spell.Transport.WebSocket`:

```elixir
def websocket_info({:send, raw_message}, _conn_state,
                   %{serializer: Serializer.JSON}) do
  {:reply, {:text, raw_message}, state}
end
def websocket_info({:send, raw_message}, _conn_state,
                   %{serializer: Serializer.MessagePack}) do
  {:reply, {:binary, raw_message}, state}
end

...

defp get_extra_headers(Serializer.JSON) do
  [{"Sec-WebSocket-Protocol", "wamp.2.json"}]
end
defp get_extra_headers(Serializer.MessagePack) do
  [{"Sec-WebSocket-Protocol", "wamp.2.msgpack"}]
end
```

That being said, none of these approaches give full extensibility -- with each, internal spell code would have to be changed for some extensions (hope that made some sense). This is probably fine for now, but something to think about.

What do you think?

**Comment by jtmoulia** [comment]
Re repeating sets of tests for the different serializers, I agree -- it would be good. Mind kicking off thinking about it by opening a new issue?

**Comment by danmarcab** [comment]
Thanks for the feedback,

I see the first approach as more extensible, but as you said both options have downsides. 
I think I'll give the first option a go and see how it looks like.
@mtanzi, @yelps what do you think?

File: lib/spell/serializer.ex [hunk 1]
```diff
@@ -8,21 +8,21 @@ defmodule Spell.Serializer do
...
 
       $transport.2.$serializer
 
   """
-  defcallback name :: String.t
+  defcallback transport_info(module :: atom) :: map
 
   @doc """
-  Set the pid which the transport should send received messages to.
+  Decodes a binary string encoded message
```
Comment (danmarcab) [inline lib/spell/serializer.ex]: This doc was confusing, I changed to something that make sense to me, but I can be missing something

Comment (jtmoulia) [inline lib/spell/serializer.ex]: Nope, you were missing nothing: looks like copy-pasta. Thanks for the catch!

**Pull request merged**